### ISSUE TITLE
Pages listing in admin area slows down the page load

### ISSUE THREAD

**Issue by acondura** [getgrav/grav#3178]
I have almost 30K pages (and many more coming) and when logging into /admin, any admin page takes 2-4 minutes to load. It used to not load at all because it would timeout after 10 min but after setting enable_auto_updates_check: false in admin.yaml, the page load dropped to 2-4min.

The next step would be to completely remove the Pages sidebar item so that the pages are not read nor counted. Ideally, the Pages listing should have a listing limit like Accounts have, 10 / listing.

The quick fix for me currently was to go override admin/themes/grav/templates/partials/nav.html.twig in a custom plugin and set:
{% if label != "Pages" %}
on line 19 as soon as the for loop start.
Unfortunately, it did not work, the Pages list item is still there and to be honest, I don't know if that will stop the pages from being read and counted when logging into admin.

Can you please shed more light on this process and what I can do about it?

**Comment by w00fz** [comment]
Are you using Grav 1.7 and Admin 1.10? The latest version feature Flex Pages and the listing does have pagination.

Or do you mean editing a page ?

**Comment by acondura** [comment]
Yes, latest 1.7.

**Comment by acondura** [comment]
I'm talking about simple admin loading, the dashboard page takes minutes for no other reason that I have 30K pages.

**Comment by acondura** [comment]
hm... checking that pagination config now.

**Comment by acondura** [comment]
I have a really long scrollbar (this is only local with almost 1400 pages, not server with 30K) and I can't find any config about pagination like in Accounts. I've enabled pages type flex in system.yaml.

![Screenshot from 2021-01-28 20-02-52](https://user-images.githubusercontent.com/1694890/106179705-146ab380-61a4-11eb-954d-8c92c02db5e4.png)

**Comment by rhukster** [comment]
YOu have 1400 pages in one level? no sub-folders? your routes indicate organization that should be handled in sub-folders.

**Comment by acondura** [comment]
True but most of the pages are tagged in other/multiple categories/paths too and I would have to have that same page in that path as well when importing from Drupal so, instead, import one page, set multiple tags in page header, build collections and cache them.

**Comment by acondura** [comment]
Even if I had sub-folders, they would still have thousands (some even tens of thousands) of pages in any given folder/sub-folder.

**Comment by acondura** [comment]
Can I disable pages completely just in admin area? I don't use them, nobody uses them, they are edited in Drupal as a backend and then simply displayed with Grav.

**Comment by mahagr** [comment]
Make sure you turn on admin caching, too. The page count is cached by default.

**Comment by acondura** [comment]
Yes, admin caching is on. Still, admin pages are painfully slow because it tries to read them (or their routes, I'm not sure what I'm seeing in debugger) on each load although I'm not accessing the Pages page or anything under that.
Why does admin work like that?

**Comment by mahagr** [comment]
Another reason may be an old plugin trying to read all the pages. Which plugins do you have installed?

**Comment by mahagr** [comment]
@acondura Maybe the easiest way to figure out what is happening is to zip the user folder and send it to me.

**Comment by acondura** [comment]
I changed jquery from 2 to 3 and it works better. Not completely sure about
this as cause since I did quite a few changes. l will check further.

**Comment by mahagr** [comment]
Anyways you can find me from Discord (Matias#1018)

**Comment by acondura** [comment]
I've updated today to 1.7.5 and then I emptied my user folder and placed in there only the stuff that comes with standard grav admin, like I would download from https://getgrav.org/download/core/grav-admin/1.7.5. However, I left the pages folder (which has 2500 pages in there) and push it to the dev server. Any admin page still loads in 7-15 seconds.

![Screenshot from 2021-02-02 16-53-01](https://user-images.githubusercontent.com/1694890/106617303-2c14b400-6577-11eb-8a87-df90fc9c6d58.png)

**Comment by acondura** [comment]
Would it help if the pages were structured like:
categ dir -> yyyymm dir -> page1, etc.
categ dir -> subcateg -> yyyymm dir -> page, etc. ?

Then, I guess I could add the additional category links (one page is in x, y and z categories) as route aliases in that page and when I'm doing a category listing of x, y or z I could probably do a taxonomy listing, right?

I guess the bottom line is if multiple subfolders would help with disk read performance?

**Comment by rhukster** [comment]
Any kind of organization is going to help things.  Having a flat structure is about the worst possible scenario for Grav (or any flat-file CMS to handle).  You are requesting every file in that folder each time.  The more you can nest/organize the better things will become. 

I suggest restructuring based on your URL structure. It would mean things are much more organized, plus you would have the benefit of having things be where you expect them to be. And best of all performance would be greatly improved.

**Comment by mahagr** [comment]
I'm also wondering how loading page index can take 2 seconds. Is the server running on a regular hard drive? Grav kind of requires SSD and enough RAM to keep index data cached in memory.

**Comment by acondura** [comment]
The user/pages folder is a symlink to persist/pages folder which is an EFS persistent volume in AWS.
I keep seeing "Page cache missed, rebuilding pages.." on any page I visit, even on local pc with docker.

**Comment by acondura** [comment]
Turns out it is a memcached connection issue because with File cache driver pages load in 2-4 seconds.

**Comment by mahagr** [comment]
Just an update; the next version will load pages much faster (when not cached) as I've done some optimization that should speed up the logic a bit.

Also, the reason why memcached was so slow is likely that it had not enough room to keep all the pages in the cache. You should make sure that the cache has enough room to keep all entries cached.

**Comment by acondura** [comment]
Great.
I ended up using only a handful of pages (less than 100 actual Grav pages objects) and the rest of hundreds of thousands of pages (Drupal nodes) up to 500K (yes, half a million) I import them as JSON files. I also had to use the functionality behind https://github.com/fbrnc/grav-plugin-staticfilecache (both the htaccess and the static html generation) so that I make sure page load stays as close as possible to 1 second.
I'm currently up to 74K imported so far and it's as fast as if there was no content to load (well, almost).
